Data/ByteString/Builder/Prim/ASCII.hs

```diff
@@ -228,64 +265,232 @@ wordHex = caseWordSize_32_64
+    :: Word32 -> Ptr Word8 -> IO (Ptr Word8)
+
+foreign import ccall unsafe "static _hs_bytestring_uint64_hex_upper" c_uint64_hex_upper
+    :: Word64 -> Ptr Word8 -> IO (Ptr Word8)
+
+{-# INLINE encodeWordHexUpper #-}
+encodeWordHexUpper :: forall a. (Storable a, Integral a) => BoundedPrim a
+encodeWordHexUpper =
+    boudedPrim (2 * sizeOf (undefined :: a)) $ c_uint32_hex_upper  . fromIntegral
+
+-- | Shortest hexadecimal encoding of a 'Word8' using upper-case characters.
+{-# INLINE word8HexUpper #-}
+word8HexUpper :: BoundedPrim Word8
+word8HexUpper = encodeWordHexUpper
+
+-- | Shortest hexadecimal encoding of a 'Word16' using upper-case characters.
+{-# INLINE word16HexUpper #-}
+word16HexUpper :: BoundedPrim Word16
+word16HexUpper = encodeWordHexUpper
+
+-- | Shortest hexadecimal encoding of a 'Word32' using upper-case characters.
+{-# INLINE word32HexUpper #-}
+word32HexUpper :: BoundedPrim Word32
+word32HexUpper = encodeWordHexUpper
+
+-- | Shortest hexadecimal encoding of a 'Word64' using upper-case characters.
+{-# INLINE word64HexUpper #-}
+word64HexUpper :: BoundedPrim Word64
+word64HexUpper = boudedPrim 16 $ c_uint64_hex_upper . fromIntegral
+
+-- | Shortest hexadecimal encoding of a 'Word' using upper-case characters.
+{-# INLINE wordHexUpper #-}
+wordHexUpper :: BoundedPrim Word
+wordHexUpper = caseWordSize_32_64
+    (fromIntegral >$< word32HexUpper)
+    (fromIntegral >$< word64HexUpper)
+
 
 -- fixed width; leading zeroes
 ------------------------------
 
--- | Encode a 'Word8' using 2 nibbles (hexadecimal digits).
-{-# INLINE word8HexFixed #-}
-word8HexFixed :: FixedPrim Word8
-word8HexFixed = fixedPrim 2 $
-    \x op -> poke (castPtr op) =<< encode8_as_16h lowerTable x
+foreign import ccall unsafe "static _hs_bytestring_builder_uint32_fixed_width_hex" c_uint32_fixed_hex
+    :: CInt -> Word32 -> Ptr Word8 -> IO ()
 
--- | Encode a 'Word16' using 4 nibbles.
-{-# INLINE word16HexFixed #-}
-word16HexFixed :: FixedPrim Word16
-word16HexFixed =
-    (\x -> (fromIntegral $ x `shiftr_w16` 8, fromIntegral x))
-      >$< pairF word8HexFixed word8HexFixed
+foreign import ccall unsafe "static _hs_bytestring_builder_uint64_fixed_width_hex" c_uint64_fixed_hex
+    :: CInt -> Word64 -> Ptr Word8 -> IO ()
 
--- | Encode a 'Word32' using 8 nibbles.
-{-# INLINE word32HexFixed #-}
-word32HexFixed :: FixedPrim Word32
-word32HexFixed =
-    (\x -> (fromIntegral $ x `shiftr_w32` 16, fromIntegral x))
-      >$< pairF word16HexFixed word16HexFixed
--- | Encode a 'Word64' using 16 nibbles.
-{-# INLINE word64HexFixed #-}
-word64HexFixed :: FixedPrim Word64
-word64HexFixed =
-    (\x -> (fromIntegral $ x `shiftr_w64` 32, fromIntegral x))
-      >$< pairF word32HexFixed word32HexFixed
+{-# INLINE encodeWordHexFixedWidth #-}
+encodeWordHexFixedWidth :: forall a. (Storable a, Integral a) => Int -> FixedPrim a
+encodeWordHexFixedWidth width = fixedPrim width $ c_uint32_fixed_hex (CInt (fromIntegral width)) . fromIntegral
```

Comment:
Should the width here be unsigned? (e.g. `Word8` rather than `Int`). Nasty surprises likely await anyone who passes a negative value, and also fixed widths above 255 bytes don't seem to have much utility.  So I think there's a case for the width being a `Word8`, but a `Word` would also be OK(ish).